ocp_resources/utils.py

```diff
@@ -26,74 +29,170 @@ class TimeoutSampler:
+
     @property
     def _func_log(self):
         return f"Function: {self.func} Kwargs: {self.func_kwargs}"
 
     def __iter__(self):
-        last_exp = None
         timeout_watch = TimeoutWatch(timeout=self.wait_timeout)
         if self.print_log:
             LOGGER.info(
                 f"Waiting for {self.wait_timeout} seconds, retry every {self.sleep} seconds"
             )
 
+        last_exp_log = "No exceptions raised"
         while timeout_watch.remaining_time() > 0:
             try:
                 self.elapsed_time = self.wait_timeout - timeout_watch.remaining_time()
                 yield self.func(**self.func_kwargs)
                 self.elapsed_time = None
+
                 time.sleep(self.sleep)
 
-            except self.exception as exp:
-                last_exp = exp
-                self.elapsed_time = None
+            except self._exceptions as exp:
+                last_exp_log = self._get_exception_log(exp=exp)
+                if self._is_raisable_exception(exp=exp):
+                    LOGGER.error(last_exp_log)
+                    raise exp
 
+                self.elapsed_time = None
                 time.sleep(self.sleep)
 
             finally:
                 if self.elapsed_time and self.print_log:
                     LOGGER.info(f"Elapsed time: {self.elapsed_time}")
 
-        raise TimeoutExpiredError(self._process_execution(exp=last_exp))
+        raise TimeoutExpiredError(last_exp_log)
 
-    def _process_execution(self, exp=None):
-        exp_name = exp.__class__.__name__ if exp else "N/A"
-        last_exception_log = f"Last exception: {exp_name}: {exp}"
+    def _message_match(self, exp, exp_class):
+        exception_messages = self.exceptions_dict.get(exp_class)
+        if not exception_messages:
+            return True
 
-        log = "{timeout}\n{func_log}\n{last_exception_log}".format(
-            timeout=self.wait_timeout,
-            func_log=self._func_log,
-            last_exception_log=last_exception_log,
-        )
+        for msg in exception_messages:
+            if msg and msg in str(exp):
+                return True
+
+        return False
+
+    def _is_raisable_exception(self, exp):
+        """
+        Verify whether exception should be raised during execution
+
+        Args:
+            exp: exception: Exception object to be tested
 
-        if self.exceptions_msg:
-            if self.exceptions_msg not in str(exp):
-                LOGGER.error(log)
-                raise exp
-            else:
-                LOGGER.warning(f"{self.exceptions_msg}: Retrying")
+        Returns:
+            bool: True if exception should be raised, False otherwise
+        """
+
+        if exp.__class__ in self.exceptions_dict:
+            if self._message_match(exp=exp, exp_class=exp.__class__):
+                return False
+
+        else:
+            for entry in self.exceptions_dict:
+                if isinstance(exp, entry) and self._message_match(
```

Comment:
This is related to the purpose of these tests, which preserve the original behavior:
https://github.com/RedHatQE/openshift-python-wrapper/pull/24#discussion_r678381822

Reply:
In fact, making exceptions_dict an `OrderedDict` would allow for a consistent exception list order, and enable priorities between different classes/messages.

Reply:
I don't think we need priorities for  exceptions, all  exceptions have the same priority